File: engines/kmp-engine/src/commonMain/kotlin/com/trajectoryruntime/engine/CatchNetworkPartition.kt
// Copyright (c) 2026 Dennis Brandl
// Licensed under the Apache License, Version 2.0. See LICENSE for details.
package com.trajectoryruntime.engine

data class CatchNetworkPartition(
    val mainFlowStepOids: Set<String>,
    val catchNetworkStepOids: Set<String>,
    /** catch_id -> set of step oids reachable from that CATCH (only if the subgraph reaches a RETURN). */
    val networksByCatchId: Map<String, Set<String>>,
)

/** Spec §6.5: a catch network is the BFS-closure from a CATCH over outgoing edges, counted only if it reaches a RETURN. */
fun partitionCatchNetworks(
    steps: List<Triple<String, String, String?>>, // (oid, stepType, catchId)
    connections: List<Pair<String, String>>,      // (fromOid, toOid)
): CatchNetworkPartition {
    val outEdges = HashMap<String, MutableList<String>>()
    for ((from, to) in connections) outEdges.getOrPut(from) { mutableListOf() }.add(to)
    val typeByOid = steps.associate { it.first to it.second }

    val networksByCatchId = HashMap<String, Set<String>>()
    val catchNetworkStepOids = HashSet<String>()
    for ((oid, type, catchId) in steps) {
        if (type != "CATCH" || catchId == null) continue
        val visited = HashSet<String>()
        val queue = ArrayDeque<String>()
        queue.add(oid)
        var hasReturn = false
        while (queue.isNotEmpty()) {
            val cur = queue.removeFirst()
            if (!visited.add(cur)) continue
            if (typeByOid[cur] == "RETURN") hasReturn = true
            for (n in outEdges[cur] ?: emptyList()) if (n !in visited) queue.add(n)
        }
        if (hasReturn) {
            networksByCatchId[catchId] = visited
            catchNetworkStepOids.addAll(visited)
        }
    }
    val mainFlowStepOids = steps.map { it.first }.filterNot { it in catchNetworkStepOids }.toSet()
    return CatchNetworkPartition(mainFlowStepOids, catchNetworkStepOids, networksByCatchId)
}

File: engines/web/src/catch-network-partition.ts
// Copyright (c) 2026 Dennis Brandl
// Licensed under the Apache License, Version 2.0. See LICENSE for details.

export interface PartitionStep { oid: string; step_type: string; catch_id?: string }
export interface PartitionConnection { from_step_id: string; to_step_id: string }

export interface CatchNetworkPartition {
  mainFlowStepOids: Set<string>;
  catchNetworkStepOids: Set<string>;
  /** catch_id → set of step oids reachable from that CATCH (only if the subgraph contains a RETURN). */
  networksByCatchId: Map<string, Set<string>>;
}

/** Spec §6.5: a catch network is the BFS-closure from a CATCH over outgoing edges, but only counts if it reaches a RETURN. */
export function partitionCatchNetworks(
  steps: PartitionStep[],
  connections: PartitionConnection[],
): CatchNetworkPartition {
  const outEdges = new Map<string, string[]>();
  for (const c of connections) {
    const list = outEdges.get(c.from_step_id);
    if (list) list.push(c.to_step_id);
    else outEdges.set(c.from_step_id, [c.to_step_id]);
  }
  const typeByOid = new Map<string, string>();
  for (const s of steps) typeByOid.set(s.oid, s.step_type);

  const networksByCatchId = new Map<string, Set<string>>();
  const catchNetworkStepOids = new Set<string>();

  for (const step of steps) {
    if (step.step_type !== 'CATCH' || !step.catch_id) continue;
    const visited = new Set<string>();
    const queue: string[] = [step.oid];
    let hasReturn = false;
    while (queue.length > 0) {
      const cur = queue.shift()!;
      if (visited.has(cur)) continue;
      visited.add(cur);
      if (typeByOid.get(cur) === 'RETURN') hasReturn = true;
      for (const n of outEdges.get(cur) ?? []) {
        if (!visited.has(n)) queue.push(n);
      }
    }
    if (hasReturn) {
      networksByCatchId.set(step.catch_id, visited);
      for (const oid of visited) catchNetworkStepOids.add(oid);
    }
  }

  const mainFlowStepOids = new Set<string>();
  for (const s of steps) {
    if (!catchNetworkStepOids.has(s.oid)) mainFlowStepOids.add(s.oid);
  }
  return { mainFlowStepOids, catchNetworkStepOids, networksByCatchId };
}
